add unused node check for compile_dag_node

LydiaXwQ opened the pull request:
## Why are these changes needed?

This will optimize DAG execution by conserving memory through the release of unused resources. Additionally, it enhances error detection by providing immediate and clear feedback on unused inputs, enabling developers to debug and refine their DAGs more effectively.
We currently store all user-accessed keys from the` input.attribute_node `into the accessed_keys, then traverse backward from the DAG’s output node to find all user input (see below). An input is considered used if a path exists from that input node to output. Finally, we compare `accessed_keys` to `used_keys`, and report errors for any `unused_keys`. Changes we did in **compiled_dag_node.py** can be quickly searched by cmd +f: “`handle_unused_keys`”.

### Code Snippet
```python
def traverse(node):
            if node in visited_nodes:
                return
            visited_nodes.add(node)

            if isinstance(node, InputAttributeNode):
                used_keys.add(node.key)

            for upstream_node in node._upstream_nodes:
                traverse(upstream_node)
        traverse(output_node)

        unused_keys = accessed_keys - used_keys

        if unused_keys:
            unused_keys_str = ', '.join(str(key) for key in unused_keys)
            accessed_keys_str = ', '.join(str(key) for key in accessed_keys)
            unused_phrase = "is unused" if len(unused_keys) == 1 else "are unused"

            raise ValueError(
                f"DAG expects input: {accessed_keys_str}, "
                f"but {unused_keys_str} {unused_phrase}. "
                "Ensure all accessed inputs from InputNode are connected to the output."
            )

```
## Related issue number
Closes #47165 

## Checks
- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

LydiaXwQ committed:
add unused node check for compile_dag_node

Signed-off-by: Lydia <wenqix3@uw.edu>

LydiaXwQ committed:
add unused node check for compile_dag_node

Signed-off-by: Lydia <wenqix3@uw.edu>

LydiaXwQ committed:
fix linter error

Signed-off-by: Lydia <wenqix3@uw.edu>

LydiaXwQ committed:
fix linter error

Signed-off-by: Lydia <wenqix3@uw.edu>

simonsays1980 committed:
[RLlib; docs] Adjust the documentation of the default API stack and replace 'rollouts' with 'env_runners'. (#49304)

angelinalg committed:
[docstring] adding link to runtime_env API ref to instances of ray_remote_args (#48955)

Per request from @GokuMohandas, add link to runtime_env API ref for all
commands with the `ray_remote_args` parameter to improve usability.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git
commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for
https://docs.ray.io/en/master/.
- [ ] I've added any new APIs to the API Reference. For example, if I
added a
method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a
few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

---------

Signed-off-by: angelinalg <angelina@anyscale.com>
Signed-off-by: angelinalg <122562471+angelinalg@users.noreply.github.com>

win5923 committed:
[Doc] Fix incorrect URL in `persist-kuberay-custom-resource-logs.md` (#49323)

Signed-off-by: win5923 <ken89@kimo.com>

dentiny committed:
[core] Minor update on task manager (#49272)

Signed-off-by: hjiang <hjiang@anyscale.com>

dentiny committed:
[core] Fix cpp inclusion check (#49301)

Signed-off-by: hjiang <dentinyhao@gmail.com>

dentiny committed:
[core] Remove unused function arg (#49331)

Signed-off-by: dentiny <dentinyhao@gmail.com>

bveeramani committed:
[Data] Set frequency of `read_images_comparison_microbenchmark_single_node` to manual (#49322)

The `read_images_comparison_microbenchmark_single_node` doesn't provide
a substantially different signal than the more-challenging `read_images`
release test. So, to save costs, this PR changes the frequency of
`read_images_comparison_microbenchmark_single_node` to manual.

Since the test could be useful if we want to compare against single-node
libraries, this PR doesn't remove the code entirely.

---------

Signed-off-by: Balaji Veeramani <bveeramani@berkeley.edu>

bveeramani committed:
[Data] Add release test for reading from URIs (#49321)

One of the most common Ray Data patterns is to load metadata containing
object (e.g., images) URIs and loading them with a map function. This PR
adds a release test to capture that use case.

---------

Signed-off-by: Balaji Veeramani <bveeramani@berkeley.edu>

dentiny committed:
[Doc] Fix doc lint (#49340)

Signed-off-by: dentiny <dentinyhao@gmail.com>

bveeramani committed:
[Data] Remove remaining push-based shuffle release tests (#49329)

As a follow-up to https://github.com/ray-project/ray/pull/48872, this PR
removes the remaining push-based shuffle release tests.

---------

Signed-off-by: Balaji Veeramani <bveeramani@berkeley.edu>

angelinalg committed:
fix indentation (#49339)

missed an indentation in this PR:
https://github.com/ray-project/ray/pull/48955

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git
commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for
https://docs.ray.io/en/master/.
- [ ] I've added any new APIs to the API Reference. For example, if I
added a
method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a
few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Signed-off-by: angelinalg <angelina@anyscale.com>

bveeramani committed:
[Data] Update iteration release tests (#49170)

The current iterations are too small-scale to capture real-world issues. To address this, this PR makes the iteration release tests multi-node and larger-scale.

---------

Signed-off-by: Balaji Veeramani <bveeramani@berkeley.edu>

bveeramani committed:
[Data] Add multi-node reading TFRecords benchmark (#49333)

To get more-useful results, this PR updates our read_tfrecords release test to use an autoscaling cluster and read a larger dataset.

---------

Signed-off-by: Balaji Veeramani <bveeramani@berkeley.edu>

alexeykudinkin committed:
[Autoscaler][Schema] Cleaned up `ClusterResourceConstraint` definition (#49311)

Signed-off-by: Alexey Kudinkin <ak@anyscale.com>

hongpeng-guo committed:
[Train][nit] Fix typo in BackendExecutor function doc str (#49347)

Fix a typo in the doc string of a function in BackendExecutor

Signed-off-by: Hongpeng Guo <hpguo@anyscale.com>

bveeramani committed:
[Data] Migrate chaos all-to-all tests (#49330)

https://github.com/ray-project/ray/pull/49017 made the sort and random
shuffle release tests us an autoscaling cluster and moved the test
definitions to `release_data_tests.yaml`. This PR is a follow-up, and
updates the chaos variants accordingly.

TODO: The random shuffle chaos test is failing. It's currently marked as
unstable so it won't block releases, but we should investigate and fix
it soon.

---------

Signed-off-by: Balaji Veeramani <bveeramani@berkeley.edu>

rynewang committed:
[core] Add call_site to Tasks and Actors. (#48920)

Adds a new field `string call_site` to TaskSpec. It's populated from
language frontend (`_raylet.pyx`), if
`RAY_record_task_actor_creation_sites` is set to `true`. The field is
propagated to `TaskEvent` and `ActorTableData` and through Dashboard
APIs.

Users can use `ray list task --detail` as well as the Web UI. It works
for Tasks, Actors (creation) and Actor Methods.

The flag `RAY_record_task_actor_creation_sites` is disabled by default,
user can enable it if they want; so by default there should be no
performance costs.


Signed-off-by: Ruiyang Wang <rywang014@gmail.com>
Signed-off-by: Ruiyang Wang <56065503+rynewang@users.noreply.github.com>
Co-authored-by: angelinalg <122562471+angelinalg@users.noreply.github.com>

dentiny committed:
[Doc] Fix doc lint (#49355)

Signed-off-by: dentiny <dentinyhao@gmail.com>

sven1977 committed:
[RLlib] Fix schedule validation on new API stack (for config settings like `lr` or `entropy_coeff`). (#49363)

akyang-anyscale committed:
[serve][bugfix] Clip number of retries shown in error message (#49318)

<!-- Thank you for your contribution! Please review
https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before
opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR.
If you don't have the access to it, we will shortly find a reviewer and
assign them to your PR. -->

## Why are these changes needed?

The status message should never display that serve will retry a
"negative" amount of times. This can happen if the retry counter is
larger than the failed threshold. If this is the case, just clip the
lower bound to be 0 to avoid the confusing status message.

## Related issue number
N/A
<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git
commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for
https://docs.ray.io/en/master/.
- [ ] I've added any new APIs to the API Reference. For example, if I
added a
method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a
few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Signed-off-by: akyang-anyscale <alexyang@anyscale.com>

sven1977 committed:
[RLlib] Cleanup examples folder vol12: GPUs on EnvRunners. (#49166)

mattip committed:
use delvewheel to repair windows wheel (#49278)

Signed-off-by: Matti Picus <matti.picus@gmail.com>

kevin85421 committed:
[core][compiled-graphs][1/N] Polish APIs: execution_timeout, retrieval_timeout, buffer_size_bytes (#49326)

Signed-off-by: kaihsun <kaihsun@anyscale.com>

jjyao committed:
[Core] Refactor ray._private.runtime_env.packaging.Protocol to make it extensible (#49341)

Signed-off-by: Jiajun Yao <jeromeyjj@gmail.com>

ruisearch42 committed:
[core] Fix doc for User-Spawned Process (#49351)

The doc was not accurate.

Signed-off-by: Rui Qiao <ruisearch42@gmail.com>

ruisearch42 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -926,6 +926,40 @@ def _add_node(self, node: "ray.dag.DAGNode") -> None:
         self.dag_node_to_idx[node] = idx
         self.counter += 1
 
+    # Helper function to handle_unused_attributes.
```
Add this as part of the docstring

ruisearch42 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -926,6 +926,40 @@ def _add_node(self, node: "ray.dag.DAGNode") -> None:
         self.counter += 1
 
+    # Helper function to handle_unused_attributes.
+    def find_unused_input_attributes(self, output_node, input_attributes) -> Set[str]:
```
type annotations for args;
the return type should be set of DAG nodes?

ruisearch42 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -926,6 +926,40 @@ def _add_node(self, node: "ray.dag.DAGNode") -> None:
+        """
+        from ray.dag import InputAttributeNode
+
+        used_attributes = set()
+        # Keep track of visited nodes during backtracking.
+        visited_nodes = set()
```
type annotations

ruisearch42 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -949,12 +983,18 @@ def _preprocess(self) -> None:
         nccl_actors_p2p: Set["ray.actor.ActorHandle"] = set()
         nccl_collective_ops: Set[_CollectiveOperation] = set()
 
+        input_attributes = set()
```
ditto: type annotation

ruisearch42 commented on python/ray/dag/tests/test_input_node.py:
```diff
@@ -379,6 +379,33 @@ def _apply_recursive_with_counter(self, fn):
+    actor = Actor.remote()
+
+    with ray.dag.InputNode() as dag_input:
+        input1, input2, input3 = dag_input[0], dag_input[1], dag_input[2]
```
nit: let's call them input0, input1 and input2

ruisearch42 commented on python/ray/dag/tests/test_input_node.py:
```diff
@@ -379,6 +379,33 @@ def _apply_recursive_with_counter(self, fn):
+        match="DAG expects input: 0, 1, 2, but 1 is unused. "
+        "Ensure all accessed inputs from InputNode are connected to the output.",
+    ):
+        _ = dag.experimental_compile()
```
you can remove `_ =`?

ruisearch42 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1129,6 +1169,22 @@ def _preprocess(self) -> None:
+            unused_phrase = "is unused" if len(unused_attributes) == 1 else "are unused"
+
+            raise ValueError(
+                f"DAG expects input: {input_attributes_str}, "
+                f"but {unused_attributes_str} {unused_phrase}. "
+                "Ensure all accessed inputs from InputNode are connected to the output."
```
Let's rephrase to:
Compiled Graph expects input to be accessed using all of attributes {input_attributes_str}, but {unused_attributes_str} {unused_phrase}. Ensure all input attributes are used and contribute to the computation of the Compiled Graph output.

ruisearch42 reviewed (commented):
Thanks for the PR! Overall looks good.

dentiny committed:
[core] Optimize health check destruction (#49231)

Signed-off-by: hjiang <hjiang@anyscale.com>

sven1977 committed:
[RLlib; docs] Docs do-over (new API stack): Env pages vol 02. (#48542)

kevin85421 committed:
[core][compiled graphs] `allreduce` errors with num returns > 1 (#49300)

Signed-off-by: Kai-Hsun Chen <kaihsun@anyscale.com>

dentiny committed:
[core] [easy] [no-op] Cleanup shared pointer usage in lease (#49362)

Signed-off-by: dentiny <dentinyhao@gmail.com>

smanolloff committed:
[RLlib] Fix incorrect "num_workers" passed in to env creator (#48540)

anyadontfly committed:
[core][compiled graphs] Add CPU-based NCCL communicator for development (#48440)

This allows developers to debug DAGs with collective ops on CPU.
Currently we use Ray actor to perform allreduce.

```python
num_workers = 2
workers = [CPUTorchTensorWorker.remote() for _ in range(num_workers)]
cpu_group = CPUCommunicator(num_workers, workers)

with InputNode() as inp:
    computes = [worker.compute.bind(inp) for worker in workers]
    collectives = collective.allreduce.bind(computes, transport=cpu_group)
    recvs = [worker.recv.bind(collective)  for worker, collective in zip(workers, collectives)]
    dag = MultiOutputNode(recvs)

compiled_dag = dag.experimental_compile()
ray.get(compiled_dag.execute())
```

Closes [#47936](https://github.com/ray-project/ray/issues/47936)
---------

Signed-off-by: tfsingh <105320310+tfsingh@users.noreply.github.com>
Co-authored-by: Tej Singh <tejfsingh@gmail.com>
Co-authored-by: tfsingh <105320310+tfsingh@users.noreply.github.com>
Co-authored-by: Puyuan Yao <puyuan@uw.edu>

hongpeng-guo committed:
[Train] Add env vars to enable Share AMD `ROCR_VISIBLE_DEVICES`  (#49346)

This PR enables to share `ROCR_VISIBLE_DEVICES` when using AMD GPUs. In
this way, the devices can see and communicate with other GPU devices.

Signed-off-by: Hongpeng Guo <hpguo@anyscale.com>

MortalHappiness committed:
[CI] Diff requirements_compiled.txt to ensure it is up to date (#49302)

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

dentiny committed:
[core] Add spdlog for rotation filename format (#49365)

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
[core] [no-op] Add doc on object recovery (#49290)

Signed-off-by: dentiny <dentinyhao@gmail.com>

dayshah committed:
[core] Memory store cv usage cleanup (#49305)

Signed-off-by: dayshah <dhyey2019@gmail.com>

dentiny committed:
[Doc] Fix lint (#49375)

Signed-off-by: dentiny <dentinyhao@gmail.com>

bveeramani committed:
[Data] Bump `batch_inference_hetero` timeout  (#49348)

https://github.com/ray-project/ray/pull/49012 decreased number of GPUs
for `batch_inference_hetero` from 40 to 10. As a result, the test has
been timeing out. This PR fixes the test by bumping up the timeout.

---------

Signed-off-by: Balaji Veeramani <bveeramani@berkeley.edu>

rynewang committed:
[core] Add cpp test for array.h (#49369)

The test is all compile time and static_assert.

---------

Signed-off-by: Ruiyang Wang <rywang014@gmail.com>

LydiaXwQ committed:
Merge branch 'master' of github.com:LydiaXwQ/ray
